### PR TITLE
Improve realtime shimming

### PR DESCRIPTION
**Pull request by po09i**
## Description
This PR improves the following

- More precise timepoints (1 per slice when possible rather than only one per volume)
- Mean pressure is taken from the average pressure in the acquisition window rather than the average pressure at the specific volume timepoints)
- New plot to visualize pressure vs field
- Modified the eval function of the Realtime sequencer to calculate the field on the original fmap (if it was changed due to having insufficient slices)
- New CLI option to change the relative time offset between the time reported in the acquisitions and the PMU

Useful ref:
- https://github.com/andrewrosss/fmri-physio-log
- https://www.cfn.upenn.edu/aguirre/wiki/doku.php?id=public:pulse-oximetry_during_fmri_scanning
- https://wiki.humanconnectome.org/display/PublicData/Understanding+Timing+Information+in+HCP+Physiological+Monitoring+Files
- https://en.wikibooks.org/wiki/SPM/Slice_Timing

### CODE REVIEW

File: shimmingtoolbox/pmu.py [hunk 1]
```diff
@@ -123,29 +123,66 @@ def read_resp(self, fname_pmu):
...
-        raster = float(self.stop_time_mdh - self.start_time_mdh) / (len(self.data)-1)
-        times = (self.start_time_mdh + raster * np.arange(len(self.data)))  # ms
+        times = self._get_all_times()
+        start_idx, stop_idx = self._get_time_indexes(start_time, stop_time)
+
+        # +1 is needed to include the stop_idx
+        return times[start_idx:stop_idx + 1]
 
+    def _get_all_times(self):
+        raster = float(self.stop_time_mdh - self.start_time_mdh) / (len(self.data) - 1)
```
Comment (jcohenadad) [inline shimmingtoolbox/pmu.py]: make sure MDH refers to the timing of the ACQUISITION (not the recon)-- add documentation in the code

File: shimming-toolbox/shimmingtoolbox/load_nifti.py [hunk 2]
```diff
@@ -12,34 +12,168 @@
 
 logger = logging.getLogger(__name__)
 PHASE_SCALING_SIEMENS = 4096
+POSSIBLE_TIMINGS = ['slice-middle', 'volume-start', 'volume-middle']
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/load_nifti.py:15]: Consider TE instead of slice-middle

File: shimming-toolbox/shimmingtoolbox/load_nifti.py [hunk 3]
```diff
@@ -12,34 +12,168 @@
...
+
+    # Start time for each volume [ms]
+    volume_start_times = np.linspace(acq_start_time_ms,
+                                     ((n_volumes - 1) * deltat_volume) + acq_start_time_ms,
+                                     n_volumes)
+
+    if when == 'volume-start':
+        return volume_start_times
+    if when == 'volume-middle':
+
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/load_nifti.py]:
```suggestion
```

File: shimming-toolbox/shimmingtoolbox/load_nifti.py [hunk 4]
```diff
@@ -12,34 +12,168 @@
...
+        Returns:
+            np.ndarray: Slice timing in ms (n_slices).
+        """
+
+        # Can be '2D' or 3D
+        mr_acquisition_type = data.get('MRAcquisitionType')
+        if mr_acquisition_type != '2D':
+            # mr_acquisition_type is None or 3D
+            logger.warning("MR acquisition type is not 2D or not set.")
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/load_nifti.py]: or not set?

File: shimming-toolbox/shimmingtoolbox/load_nifti.py [hunk 5]
```diff
@@ -12,34 +12,168 @@
...
+
+        # If the slice timing is lower than the TR excitation, this is an interleaved multi-slice acquisition
+        # (more than one slice acquired within a TR excitation)
+        repetition_slice_excitation = repetition_slice_excitation * 1000  # [ms]
+        # Remove slices that are at 0 ms (acquired during the first TR excitation)
+        slice_timing_start_no_zero = slice_timing_start[slice_timing_start > 0]
+        # If there are other slices acquired before the next TR excitation, then this is interleaved
+        # multi-slice
+        if np.any(slice_timing_start_no_zero < repetition_slice_excitation):
+            logger.warning("Interleaved multi-slice acquisition detected.")
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/load_nifti.py:102]: I would add what is being returned. 
`Interleaved multi-slice acquisition detected, returning 0s` or sonething similar

File: shimming-toolbox/shimmingtoolbox/load_nifti.py [hunk 6]
```diff
@@ -12,34 +12,168 @@
...
+            return np.zeros(n_slices)
+
+        pulse_sequence_details = data.get('PulseSequenceDetails')
+        if pulse_sequence_details is None:
+            logger.warning("No PulseSequenceDetails name found in JSON file.")
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/load_nifti.py:107]: idem

File: shimming-toolbox/shimmingtoolbox/load_nifti.py [hunk 7]
```diff
@@ -12,34 +12,168 @@
...
+            # This protocol acquires 2 echos with 2 different TRs
+            deltat_slice = repetition_slice_excitation * phase_encode_steps * 2
+        elif "%SiemensSeq%\\gre" == pulse_sequence_details or "gre_PMUlog" in pulse_sequence_details:
+            deltat_slice = repetition_slice_excitation * phase_encode_steps
+        else:
+            logger.warning("Protocol name not recognized.")
+            return np.zeros(n_slices)
+
+        # Error check
+        volume_tr = data.get('RepetitionTime')
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/load_nifti.py:120]: What is the difference with deltat_vol

File: shimming-toolbox/shimmingtoolbox/load_nifti.py [hunk 8]
```diff
@@ -12,34 +12,168 @@
...
+                fourier = 1.0
+            else:
+                fourier = float(fourier)
+                if not (0 <= fourier <= 1):
+                    logger.warning("Partial Fourier value format not supported, make sure it is between 0 and 1.")
+                    return np.zeros(n_slices)
+
+            # When partial fourier is used, crossing of the middle of k-space is shifted
+            # For Siemens, it seems to start with the "smaller" side
+            ratio = (fourier - 0.5) / fourier
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/load_nifti.py:152]: Unclear. If partial fourier is 7/8 --> 7/8 - 4/8 = 3/8 --> 3/8 / 7/8 = 3/7.
Why 3/7 shift?

File: shimming-toolbox/shimmingtoolbox/pmu.py [hunk 9]
```diff
@@ -110,9 +118,19 @@ def read_resp(self, fname_pmu):
         # That is we assume they do *not* occupy a raster position.
         data_cleaned = data[data < 4096]
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/pmu.py:156]: Isn't 4096 the max value the scanner gives?

File: shimming-toolbox/shimmingtoolbox/pmu.py [hunk 10]
```diff
@@ -123,34 +141,160 @@ def read_resp(self, fname_pmu):
...
+        Offset the start and end time of the PMU data
+        Args:
+            time_offset (int): Time offset in ms to what is read in the .resp file
+
+        """
+        self.start_time_mdh += time_offset - self.time_offset
+        self.stop_time_mdh += time_offset - self.time_offset
+        self.start_time_mpcu += time_offset - self.time_offset
+        self.stop_time_mpcu += time_offset - self.time_offset
+        self.time_offset = time_offset
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/pmu.py]: This is confusing, you should change the names to better explain the difference between time_offset and self.time_offset

File: shimming-toolbox/shimmingtoolbox/pmu.py [hunk 11]
```diff
@@ -123,34 +141,160 @@ def read_resp(self, fname_pmu):
...
+            start_time (int): Start time in milliseconds past midnight
+            stop_time (int): Stop time in milliseconds past midnight
+
         Returns:
             np.ndarray: Array containing the timepoints in ms of each data
         """
-        raster = float(self.stop_time_mdh - self.start_time_mdh) / (len(self.data)-1)
-        times = (self.start_time_mdh + raster * np.arange(len(self.data)))  # ms
+        times = self._get_all_times()
+        start_idx, stop_idx = self._get_time_indexes(start_time, stop_time)
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/pmu.py]: times is recalculated in self._get_time_indexes

File: shimming-toolbox/shimmingtoolbox/pmu.py [hunk 12]
```diff
@@ -123,34 +141,160 @@ def read_resp(self, fname_pmu):
...
+
+        return times[start_idx:stop_idx + 1]
 
+    def _get_all_times(self):
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/pmu.py]: missing description

File: shimming-toolbox/shimmingtoolbox/pmu.py [hunk 13]
```diff
@@ -123,34 +141,160 @@ def read_resp(self, fname_pmu):
...
+        raster = float(self.stop_time_mdh - self.start_time_mdh) / (len(self.data) - 1)
+        times = (self.start_time_mdh + raster * np.arange(len(self.data)))  # ms
         return times
 
+    def _get_time_indexes(self, start_time=None, stop_time=None):
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/pmu.py:219]: missing description

File: shimming-toolbox/shimmingtoolbox/pmu.py [hunk 14]
```diff
@@ -123,34 +141,160 @@ def read_resp(self, fname_pmu):
...
+            acquisition_times (numpy.ndarray): Array of the times in milliseconds past midnight of the desired
                                                times to interpolate the resp_trace. Times must be within
                                                ``self.start_time_mdh`` and ``self.stop_time_mdh``
 
         Returns:
-            numpy.ndarray: 1D array with interpolated times
+            numpy.ndarray: Array with interpolated times with the same shape as ``acquisition_times``
         """
+        start_offset = np.min(acquisition_times - self.start_time_mdh)
+        stop_offset = np.min(self.stop_time_mdh - acquisition_times)
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/pmu.py]: Put those lines inside the if statement as they are not useful otherwise

File: shimming-toolbox/shimmingtoolbox/pmu.py [hunk 15]
```diff
@@ -123,34 +141,160 @@ def read_resp(self, fname_pmu):
...
+
+        Args:
+            start_time (int): Start time in milliseconds past midnight
+            stop_time (int): Stop time in milliseconds past midnight
+
+        Returns:
+            float: Mean value of the resp trace between ``start_time`` and ``stop_time``
+        """
+
+        pressures = self.get_resp_trace(start_time=start_time, stop_time=stop_time)
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/pmu.py:294]: many functions like get_resp_trace() and get_all_times() seem to be used a lot. It might be beneficial call these functions in the constructor and store their values in a class variable.

File: shimming-toolbox/shimmingtoolbox/shim/sequencer.py [hunk 16]
```diff
@@ -1148,18 +1151,161 @@ def __init__(self, nii_fieldmap, json_fmap, nii_anat, nii_static_mask, nii_riro_
...
+        # Todo: deal with loads of warnings when setting the time offset
+        # Probably sweep at all times but centered on the frequency added
+        n_slices = self.nii_fieldmap_orig.shape[2]
+
+        previous_time_offset = self.pmu.time_offset
+        self.pmu.adjust_start_time(0)
+        mean_respiratory_cycle_time = self.pmu.get_mean_trigger_span() / 2
+        acq_times = get_acquisition_times(self.nii_fieldmap_orig, self.json_fmap, when='slice-middle')
+        n_samples = 1000
+        min_offset = max(-mean_respiratory_cycle_time / 2, self.pmu.start_time_mdh - acq_times.min())
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/shim/sequencer.py]: Divide by 2 a second time?

File: shimming-toolbox/shimmingtoolbox/shim/sequencer.py [hunk 17]
```diff
@@ -1148,18 +1151,161 @@ def __init__(self, nii_fieldmap, json_fmap, nii_anat, nii_static_mask, nii_riro_
...
+        max_offset = min(mean_respiratory_cycle_time / 2, self.pmu.stop_time_mdh - acq_times.max())
+        # min_offset = -mean_respiratory_cycle_time / 2
+        # max_offset = mean_respiratory_cycle_time / 2
+        time_offsets = np.linspace(min_offset, max_offset, n_samples)
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/shim/sequencer.py]: min and max_offset names induce error in thinking they are the minimum and maximum offsets found

File: shimming-toolbox/shimmingtoolbox/shim/sequencer.py [hunk 18]
```diff
@@ -1148,18 +1151,161 @@ def __init__(self, nii_fieldmap, json_fmap, nii_anat, nii_static_mask, nii_riro_
...
+        best_r2total = 0
+        best_time_offset = 0
+        r2total_list = []
+        for time_offset in time_offsets:
+            self.pmu.adjust_start_time(round(time_offset))
+            r2_total = 0
+            for i_slice in range(n_slices):
+                if i_slice in []:
+                    continue
+                pressures = self.pmu.interp_resp_trace(acq_times) - 2048
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/shim/sequencer.py]: maybe avoid hardcoding 4096 and 2048.

File: shimming-toolbox/shimmingtoolbox/shim/sequencer.py [hunk 19]
```diff
@@ -1148,18 +1151,161 @@ def __init__(self, nii_fieldmap, json_fmap, nii_anat, nii_static_mask, nii_riro_
...
+        time_offsets = np.linspace(min_offset, max_offset, n_samples)
+
+        mask_fmap = np.logical_or(self.mask_static_orig_fmcs, self.mask_riro_orig_fmcs)
+        mask_4d = np.repeat(np.expand_dims(mask_fmap, axis=-1), self.nii_fieldmap_orig.shape[-1], axis=-1)
+        fmap_ma = np.ma.array(self.nii_fieldmap_orig.get_fdata(), mask=mask_4d == False)
+
+        # Find best time offset
+        best_r2total = 0
+        best_time_offset = 0
+        r2total_list = []
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/shim/sequencer.py]:
```suggestion
        r2_total_list = []
```

File: shimming-toolbox/shimmingtoolbox/shim/sequencer.py [hunk 20]
```diff
@@ -1148,18 +1151,161 @@ def __init__(self, nii_fieldmap, json_fmap, nii_anat, nii_static_mask, nii_riro_
...
+                pressures = self.pmu.interp_resp_trace(acq_times) - 2048
+                y = fmap_ma.mean(axis=(0, 1))[i_slice].filled()
+                y = (y - y.mean())
+                reg = LinearRegression().fit(pressures[:, i_slice].reshape(-1, 1), y)
+                # Adjusted r2 score
+                r2 = reg.score(pressures[:, i_slice].reshape(-1, 1), y)
+                # r2_corr = (1 - (1 - r2) * (len(y) - 1) / (len(y) - pressures[:, i_slice].reshape(-1, 1).shape[1] - 1))
+                r2_total += r2
+            r2_total /= n_slices
+            r2total_list.append(r2_total)
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/shim/sequencer.py]:
```suggestion
            r2_total_list.append(r2_total)
```

File: shimming-toolbox/shimmingtoolbox/shim/sequencer.py [hunk 21]
```diff
@@ -1148,18 +1151,161 @@ def __init__(self, nii_fieldmap, json_fmap, nii_anat, nii_static_mask, nii_riro_
...
+                best_r2total = r2_total
+                best_time_offset = time_offset
+
+        logger.info(f"Best time offset: {round(best_time_offset)}ms")
+        logger.info(f"Average r2 score: {best_r2total} at this time offset")
+
+        if self.path_output is not None:
+            fig = Figure(figsize=(8, 10))
+            ax1 = fig.add_subplot(311)
+            ax1.plot(time_offsets, r2total_list)
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/shim/sequencer.py]:
```suggestion
            ax1.plot(time_offsets, r2_total_list)
```

File: shimming-toolbox/shimmingtoolbox/shim/sequencer.py [hunk 22]
```diff
@@ -1148,18 +1151,161 @@ def __init__(self, nii_fieldmap, json_fmap, nii_anat, nii_static_mask, nii_riro_
...
+            ax1.set_xlabel("Time offset [ms]")
+            ax1.set_ylabel("Average r2")
+            ax1.set_title("R2 score for different time offsets")
+            ax1.set_ylim([-0.1, 1.1])
+
+            # 750 ms is chosen as the smoothing length
+            window_length = round(750 / ((max_offset - min_offset) / n_samples))
+            r2_list_smooth = savgol_filter(r2total_list, window_length, 4, mode='mirror')
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/shim/sequencer.py]:
```suggestion
            r2_list_smooth = savgol_filter(r2_total_list, window_length, 4, mode='mirror')
```

File: shimming-toolbox/shimmingtoolbox/shim/sequencer.py [hunk 23]
```diff
@@ -1148,18 +1151,161 @@ def __init__(self, nii_fieldmap, json_fmap, nii_anat, nii_static_mask, nii_riro_
...
+            # The distance parameter is the minimum number of samples between adjacent peaks
+            # 1000 ms is chosen as the minimum time between peaks
+            min_distance = round(500 / ((max_offset - min_offset) / n_samples))
+            peak_indices = find_peaks(r2_list_smooth, distance=min_distance, height=0.4)[0]
+            for index in peak_indices:
+                ax1.vlines(time_offsets[index], -1, 2, colors='k', linestyles='dashed')
+                ax1.annotate(f"{round(time_offsets[index])}ms",
+                             (time_offsets[index] + 50, r2total_list[index] - 0.2))
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/shim/sequencer.py]:
```suggestion
                             (time_offsets[index] + 50, r2_total_list[index] - 0.2))
```

File: shimming-toolbox/shimmingtoolbox/shim/sequencer.py [hunk 24]
```diff
@@ -1628,88 +1840,95 @@ def plot_currents(self, static, riro=None):
...
+        plots = []
+        for i_shim in range(n_shims):
+            if np.any(mask_fm[..., i_shim] != 0):
+                plots.append(i_shim)
+
+        path_pressure_and_unshimmed_field = os.path.join(self.path_output, 'fig_noshim_vs_pressure_regression')
+        create_output_dir(path_pressure_and_unshimmed_field)
+
+        for i_plot, i_shim in enumerate(plots):
+
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/shim/sequencer.py]:
```suggestion
```

File: shimming-toolbox/shimmingtoolbox/load_nifti.py (hunk 2 above)
Comment (po09i) [reply shimming-toolbox/shimmingtoolbox/load_nifti.py:15]: This is also for GRE acquisitions, so TE is not appropriate in that case...

File: shimming-toolbox/shimmingtoolbox/load_nifti.py (hunk 4 above)
Comment (po09i) [reply shimming-toolbox/shimmingtoolbox/load_nifti.py]: If it's None (tag is not present in the sidecar)

File: shimming-toolbox/shimmingtoolbox/load_nifti.py (hunk 5 above)
Comment (po09i) [reply shimming-toolbox/shimmingtoolbox/load_nifti.py:102]: This is more a user warning. A second warning is displayed later when it detects that the output of this function is all 0s mentioning what will be done instead ("Could not figure out the slice timing. Using one time-point per volume instead.")

File: shimming-toolbox/shimmingtoolbox/load_nifti.py (hunk 7 above)
Comment (po09i) [reply shimming-toolbox/shimmingtoolbox/load_nifti.py:120]: It's a string that is in seconds. It also could be None if the tag is not present.

File: shimming-toolbox/shimmingtoolbox/load_nifti.py (hunk 8 above)
Comment (po09i) [reply shimming-toolbox/shimmingtoolbox/load_nifti.py:152]: The crossing of the center-line of k-space in the phase encoding direction is not exactly at 1/2 of the time it takes to acquire a slice if partial Fourier is not 1. In your example, a 7/8 partial Fourier would acquire the center-line of k-space after 3/7 of the time it takes to acquire a slice. (For a 700ms slice, the center-line would be acquired ~300ms). I simulated a couple of sequences in POET and observed that the "smaller" portion is always acquired first (the 1/8 portion is skipped, then 3/8 is acquired, k-space is crossed then the final 1/2 is acquired. In total, the k-space crossing was at 3/7 of the slice time).

File: shimming-toolbox/shimmingtoolbox/pmu.py (hunk 9 above)
Comment (po09i) [reply shimming-toolbox/shimmingtoolbox/pmu.py:156]: 4095 is the maximum value that the pressure can have. There can be other values throughout the data that mean other things. I believe it can mean that a trigger is there (when the scanner detect a inspiration to expiration for example).

File: shimming-toolbox/shimmingtoolbox/pmu.py (hunk 15 above)
Comment (po09i) [reply shimming-toolbox/shimmingtoolbox/pmu.py:294]: I don't see a big advantage of doing it that way. For example get_resp_trace is essentially only slicing the whole respiratory data depending on the time of the input. get_all_times() would make more sense but I don't think it's really worth a refactor for the sub-millisecond time-save.

File: shimming-toolbox/shimmingtoolbox/pmu.py (hunk 10 above)
Comment (po09i) [reply shimming-toolbox/shimmingtoolbox/pmu.py]: Fixed [ae05208](https://github.com/shimming-toolbox/shimming-toolbox/pull/482/commits/ae052085f4e07165f9329c51d4c59d77910adb20)

File: shimming-toolbox/shimmingtoolbox/pmu.py (hunk 11 above)
Comment (po09i) [reply shimming-toolbox/shimmingtoolbox/pmu.py]: I moved it after the recalculation: [ae05208](https://github.com/shimming-toolbox/shimming-toolbox/pull/482/commits/ae052085f4e07165f9329c51d4c59d77910adb20)

File: shimming-toolbox/shimmingtoolbox/pmu.py (hunk 12 above)
Comment (po09i) [reply shimming-toolbox/shimmingtoolbox/pmu.py]: Fixed [ae05208](https://github.com/shimming-toolbox/shimming-toolbox/pull/482/commits/ae052085f4e07165f9329c51d4c59d77910adb20)

File: shimming-toolbox/shimmingtoolbox/pmu.py (hunk 14 above)
Comment (po09i) [reply shimming-toolbox/shimmingtoolbox/pmu.py]: Indeed [ae05208](https://github.com/shimming-toolbox/shimming-toolbox/pull/482/commits/ae052085f4e07165f9329c51d4c59d77910adb20)

File: shimming-toolbox/shimmingtoolbox/load_nifti.py (hunk 4 above)
Comment (4rnaudB) [reply shimming-toolbox/shimmingtoolbox/load_nifti.py]: Maybe rephrase, it sounds off

File: shimming-toolbox/shimmingtoolbox/load_nifti.py (hunk 5 above)
Comment (4rnaudB) [reply shimming-toolbox/shimmingtoolbox/load_nifti.py:102]: I would still add it here so the user understand that Interleaved multi-slice returns all zeros

File: shimming-toolbox/shimmingtoolbox/load_nifti.py (hunk 7 above)
Comment (4rnaudB) [reply shimming-toolbox/shimmingtoolbox/load_nifti.py:120]: But both are the 'RepetitionTime' from the json file

File: shimming-toolbox/shimmingtoolbox/load_nifti.py (hunk 8 above)
Comment (4rnaudB) [reply shimming-toolbox/shimmingtoolbox/load_nifti.py:152]: But with the explanation provided `# For Siemens, it seems to start with the "smaller" side`, I would expect the ration to be 3/8 (3/8 being the small side and 4/8 the big side). Maybe it is just me, but I would like a better explanation for the future.

File: shimming-toolbox/shimmingtoolbox/pmu.py (hunk 11 above)
Comment (4rnaudB) [reply shimming-toolbox/shimmingtoolbox/pmu.py]: It is still computed twice

File: shimming-toolbox/shimmingtoolbox/pmu.py (hunk 15 above)
Comment (4rnaudB) [reply shimming-toolbox/shimmingtoolbox/pmu.py:294]: It is not really for time efficiency, but more for overall code. It would look cleaner to save the output in the constructor and call it as a variable afterwards, since the exact same function is called with the exact same parameters. It is a very small refactoring.

File: shimming-toolbox/shimmingtoolbox/load_nifti.py (hunk 5 above)
Comment (po09i) [reply shimming-toolbox/shimmingtoolbox/load_nifti.py:102]: I'll use errors and a "try except"

File: shimming-toolbox/shimmingtoolbox/load_nifti.py (hunk 8 above)
Comment (po09i) [reply shimming-toolbox/shimmingtoolbox/load_nifti.py:152]: I will write the example in the code for the future.

Add error if < 1/2

File: shimming-toolbox/shimmingtoolbox/pmu.py (hunk 11 above)
Comment (po09i) [reply shimming-toolbox/shimmingtoolbox/pmu.py]: I'll create a variable called "timepoints"

File: shimming-toolbox/shimmingtoolbox/shim/sequencer.py (hunk 16 above)
Comment (po09i) [reply shimming-toolbox/shimmingtoolbox/shim/sequencer.py]: Variable name should be changed

File: shimming-toolbox/shimmingtoolbox/load_nifti.py [hunk 25]
```diff
@@ -12,34 +12,167 @@
...
     # Get number of volumes
     n_volumes = nii_data.header['dim'][4]
+    n_slices = nii_data.shape[2]
+
+    # Time between the beginning of the acquisition of a volume and the beginning of the acquisition of the next volume
+    deltat_volume = float(json_data['RepetitionTime']) * 1000  # [ms]
+
+    # Time the acquisition of data for this image started (ISO format)
+    acq_start_time_iso = json_data['AcquisitionTime']
+    # todo: dummy scans?
```
Comment (4rnaudB) [inline shimming-toolbox/shimmingtoolbox/load_nifti.py:42]: I don't think dummy scan affects the start time, but it needs to be tested somehow

**Review by 4rnaudB — APPROVED** [review]
